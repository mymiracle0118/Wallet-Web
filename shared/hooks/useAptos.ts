import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
