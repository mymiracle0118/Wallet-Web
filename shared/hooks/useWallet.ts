import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { ethers } from 'ethers'
import produce from 'immer'

import { default as defaultAssets } from '../data/assets.json'
import { default as networkTokens } from '../data/networkTokens.json'
import { default as networkUrls } from '../data/networkUrls.json'

import moment from 'moment'

import { toUTF8Array } from '@portal/shared/utils/utf8'
import { getNetworkProvider } from '@portal/shared/utils/getNetworkProvider'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { ContractTransaction } from '@portal/shared/services/etherscan'
import { NetworkFactory } from '../factory/network.factory'

import { useSettings } from './useSettings'
import { EnvironmentType, GasOption, NetworkToken, NetworkTokensList } from '@portal/shared/utils/types'
// @ts-ignore
import defaultAvatar from 'assets/images/Avatar.png'
import { useStore } from './useStore'
import { decryptData, encryptData } from '@portal/portal-extension/src/utils/constants'

export type AssetType = 'layer1' | 'erc20' | 'erc721'

export type listNetwork = {
  image?: string
  coin?: string
  name: string
  symbol?: string
  chainId?: number
  chain: string
  testNetwork?: boolean
  networkId: string
  blockExplorer?: string
  enabledAsDefault?: boolean
  networkURL?: string
}

export type Asset = {
  id: string
  network: string
  chain: 'ethereum' | string
  type: AssetType
  contractAddress?: string
  name: string
  symbol?: string
  decimal?: ethers.BigNumberish
  balance?: ethers.BigNumberish
  formattedBalance?: number
  token?: string
  image?: string
  metadata?: { image: string; attributes: Array<{ trait_type: string; value: string }> }
  enabledAsDefault?: boolean
  gasSymbol?: string
  isFavorite?: boolean
}

export type Transaction = {
  symbol: any
  asset: NetworkToken
  to: string
  amount?: string
  gasOption: GasOption
  hash?: string
  nonce?: number
}

export type NetworkAssetsCollection = {
  networkId: string
  coin: string
  image?: string
  imageURL?: string
  address?: string
  active?: boolean
  link?: boolean
  testNetwork?: boolean
  symbol?: string
  name?: string
  chainId?: number
  chain?: string
  rpcUrl?: string
  networkURL?: string
  blockExplorer?: string
  enabledAsDefault?: boolean
}

export type TokenProfitCollection = {
  network: string
  token: string
  value: number
  time: Date
}

export type WalletState = {
  isAccountCreatedByPrivateKey: boolean
  isAccountImported: boolean
  assets: Array<Asset>
  NFTs: Array<Asset>
  networkAssets: Array<NetworkAssetsCollection>
  address?: string
  username?: string
  avatar?: string
  encryptedWallet?: string
  wallet?: ethers.Wallet | any
  encryptedPrivateKey?: (string | number)[] | undefined
  transaction?: Transaction
  transactionDetails?: ContractTransaction
  pendingTransactions?: Array<ContractTransaction>
  lockTime: number
  lockWallet: boolean
  createWallet: () => void
  clearWallet: () => void
  suiWallet?: any
  setSuiWallet?: (wallet: any) => void
  storeWallet: (
    walletObj: any,
    username: string,
    password: string,
    network?: string,
    accountKey?: string,
    deriveIndex?: number
  ) => Promise<string>
  openWallet: (password: string) => Promise<void>
  recoverWallet: (mnemonic: string) => void
  getAsset: (network: string, assetId: string) => Asset
  getDefaultAsset: (network: string, assetId: string) => Asset
  getNFTAsset: (network: string, contractAddress: string, assetId: string) => Asset
  setTransaction: (transaction: Transaction) => void
  addNetworkAsset: (networkAsset: NetworkAssetsCollection) => void
  removeNetworkAsset: (networkAssetId: string) => void
  setTransactionDetails: (transactionDetails: ContractTransaction) => void
  setPendingTransactions: (pendingTransactions: ContractTransaction[] | undefined) => void
  sendTransaction: (
    maxPriorityFeePerGas: ethers.BigNumber,
    maxFeePerGas: ethers.BigNumber,
    gasLimit: number
  ) => Promise<{}>
  cancelTransaction: (
    network: string,
    txData: {
      nonce: string
      maxPriorityFeePerGas: ethers.BigNumber
      maxFeePerGas: ethers.BigNumber
    }
  ) => Promise<void>
  setLockTimer: (lockTime: number) => void
  setLockWallet: (lock: boolean) => void
  setActiveWallet: (address: string) => void
  addCustomToken: (tokenData: Asset) => Promise<void>
  removeToken: (network: string, assetId: string) => void
  addNFT: (nft: Asset) => void
  removeNFT: (nft: Asset) => void
  changeAvatar: (avatar: string) => void
  getNetworkList: () => Array<{ network: string; image: string | undefined }>
  tokenProfitLossCollection: Array<TokenProfitCollection>
  setTokenProfitLoss: (data: TokenProfitCollection) => void
  setUsername: (username: string) => void
  selectedNetwork?: string
  setSelectedNetwork: (selectedNetwork: string) => void
  networkType: string
  setNetworkType: (type: any) => void
  defaultOnboardingNetwork: string
  defaultNetworkTokens: { [key: string]: { [x: string]: string } } | null
  defaultNetworkUrls: { [key: string]: { [envType: string]: { [key: string]: string } } } | null
  walletAddress: { [key: string]: { [key: string]: string } }
  tokenArrayWithBalance: NetworkTokensList
  currentSelectedToken: NetworkToken | null
  selectedTokensList: string[]
  tokensList: NetworkTokensList
  getNetworksList: () => NetworkToken[]
  getNetworksTokenList: () => NetworkTokensList
  refreshTokenList: (envType: EnvironmentType) => void
  getNetworkTokenWithCurrentEnv: (key: string) => NetworkToken
}

export type AccountWallet = {
  address: string
  encryptedWallet?: string | null
  encryptedPrivateKey?: (string | number)[] | undefined
  isPrimary: boolean
  deriveIndex?: number
}

const getTokensListWithEnvUrls = (envType: EnvironmentType): NetworkTokensList => {
  let tokens: NetworkTokensList = {}
  for (const key in networkTokens) {
    const token = networkTokens[key]
    tokens[key] = { ...token, ...networkUrls[token.networkName][envType] }
  }
  return tokens
}
const initialState = {
  assets: [],
  networkAssets: [],
  address: undefined,
  username: undefined,
  avatar: undefined,
  encryptedWallet: undefined,
  wallet: undefined,
  NFTs: [],
  lockTime: 5,
  lockWallet: false,
  pendingTransactions: [],
  tokenProfitLossCollection: [],
  isAccountImported: false,
  selectedNetwork: '',
  networkType: 'testnet',
  isAccountCreatedByPrivateKey: false,
  defaultOnboardingNetwork: 'ETH',
  defaultNetworkTokens: null, // networkTokens as any,
  defaultNetworkUrls: null, //networkUrls as any,
  walletAddress: {},
  tokenArrayWithBalance: {},
  currentSelectedToken: null,
  selectedTokensList: [],
  tokensList: getTokensListWithEnvUrls('testNet'),
}

const encryptOptions = {
  scrypt: {
    N: 1 << 16,
  },
}

export const defaultNetworkTokens = networkTokens as any
export const defaultNetworkUrls = networkUrls as any

export const useWallet = create<WalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,
        refreshTokenList: (envType: EnvironmentType) => {
          const tokens: { [key: string]: NetworkToken } = getTokensListWithEnvUrls(envType)
          set({ tokensList: tokens })
        },
        getNetworkTokenWithCurrentEnv: (key: string): NetworkToken => {
          const { networkEnvironment } = useSettings.getState()
          const networkTokenList = get().getNetworksTokenList()
          const token = networkTokenList[key] as NetworkToken
          if (token) {
            const networkUrlsObject =
              networkUrls[token.networkName] && networkUrls[token.networkName][networkEnvironment]
                ? { ...networkUrls[token.networkName][networkEnvironment] }
                : {}
            return { ...token, ...networkUrlsObject }
          }
          throw new Error(`Invalid network token :: ${key}`)
        },
        getNetworksTokenList: (): NetworkTokensList | any => {
          const { networkEnvironment, currentAccount, accounts } = useSettings.getState()

          const defaultNetworkTokensList = networkTokens as unknown as NetworkTokensList
          let accountsCustomToken: NetworkTokensList =
            currentAccount && accounts[currentAccount.address]
              ? accounts[currentAccount.address].customTokens[networkEnvironment]
              : {}
          return { ...defaultNetworkTokensList, ...accountsCustomToken }
        },
        getNetworksList: (): NetworkToken[] => {
          const { tokensList } = get()
          const networkTokenList = get().getNetworksTokenList()
          const tokens: NetworkToken[] = Object.values(networkTokenList).filter((token) => {
            return token.tokenType === 'Native'
          }) as NetworkToken[]
          return tokens
        },
        setNetworkType: (type) => {
          set({ networkType: type })
        },
        createWallet: () => {
          const asset = defaultAssets.find((n) => n.network === 'mainnet') as Asset
          const wallet = NetworkFactory.selectNetwork(asset).createWallet()
          set(wallet)
        },
        setSuiWallet: (wallet) => {
          set({ suiWallet: wallet })
        },
        clearWallet: () => {
          set(initialState)
          useStore.getState().clearStore()
        },

        addNetworkAsset: (networkAsset: NetworkAssetsCollection) => {
          set(
            produce((state: WalletState) => {
              state.networkAssets.push(networkAsset)
            })
          )
        },
        removeNetworkAsset: (networkAssetId: string) => {
          set(
            produce((state: WalletState) => {
              state.networkAssets = state.networkAssets.filter((n) => n.networkId !== networkAssetId)
            })
          )
        },
        storeWallet: async (
          walletObj: any,
          username: string,
          password: string,
          network: string,
          accountKey?: string
        ) => {
          const { wallet, encryptedPrivateKey, address, deriveIndex } = walletObj
          const networkToken = get().getNetworkTokenWithCurrentEnv(network)

          if (!wallet) {
            throw new Error('no wallet created')
          }

          const walletAddress = address ? address : wallet.address
          let encryptedWallet = null

          if (networkToken.isEVMNetwork) {
            encryptedWallet = await wallet.encrypt(password, encryptOptions)
          } else {
            encryptedWallet = encryptData(wallet, password)
          }

          const newWalletObj: AccountWallet = {
            address: walletAddress,
            encryptedPrivateKey: encryptedPrivateKey,
            encryptedWallet,
            isPrimary: false,
            deriveIndex: deriveIndex,
          }
          const { addWalletToList, addTokenToList } = useStore.getState()
          await addWalletToList(accountKey || walletAddress, networkToken.networkName, newWalletObj)
          await addTokenToList(accountKey || walletAddress, networkToken.shortName)

          return walletAddress
        },

        setActiveWallet: async (address: string) => {
          const accounts = useSettings.getState().accounts
          const account = accounts[address]
          if (account) {
            set({ ...account })
          }
        },

        openWallet: async (password: string) => {
          const { currentAccount } = useSettings.getState()
          const { walletsList } = useStore.getState()
          try {
            if (currentAccount) {
              const accountWallet = walletsList[currentAccount.address][currentAccount.networkName]
              const networkToken = useWallet.getState().getNetworkTokenWithCurrentEnv(currentAccount.networkName)
              let checkPassword: any
              if (networkToken.isEVMNetwork) {
                checkPassword = await ethers.Wallet.fromEncryptedJson(
                  accountWallet.encryptedWallet as string,
                  password as string
                )
              } else {
                checkPassword = await decryptData(accountWallet.encryptedWallet as string, password as string)
              }
              if (checkPassword) {
                set({
                  lockWallet: false,
                })
              }
            }
            throw new Error('Invalid password.')
          } catch (error) {
            throw new Error(`Error: ${error.message}`)
          }
        },

        recoverWallet: (mnemonic: string) => {
          if (!ethers.utils.isValidMnemonic(mnemonic)) {
            throw new Error('Incorrect secret recovery phrase')
          }

          const wallet = ethers.Wallet.fromMnemonic(mnemonic)
          set({ wallet, encryptedPrivateKey: toUTF8Array(wallet.privateKey) })
        },

        getAsset: (network: string, assetId: string) => {
          const { assets } = get()
          const asset = assets.find((a) => a.id === assetId && a.network === network)
          if (!asset) {
            throw new Error(`Asset ${assetId} on ${network} network not found`)
          }
          return asset
        },

        getDefaultAsset: (network: string, assetId: string) => {
          const defaultAsset: Asset | undefined = (defaultAssets as Asset[]).find(
            (a) => a.id === assetId && a.network === network
          )
          if (!defaultAsset) {
            throw new Error(`Asset ${assetId} on ${network} network not found`)
          }
          return defaultAsset
        },

        getNFTAsset: (network: string, contractAddress: string, assetId: string) => {
          if (network === 'APTOS') {
            throw new Error(`NFT Asset ${assetId} on ${network} network not found`)
          }
          const { NFTs } = get()
          const asset = NFTs.find(
            (a) => a.id === assetId && a.contractAddress === contractAddress && a.network === network
          )
          if (!asset) {
            throw new Error(`NFT Asset ${assetId} on ${network} network not found`)
          }
          return asset
        },

        setTransaction: (transaction: Transaction) => {
          set({ transaction })
        },

        setTransactionDetails: (transactionDetails: ContractTransaction) => {
          set({ transactionDetails })
        },

        setPendingTransactions: (pendingTransactions: ContractTransaction[] | undefined) => {
          set({ pendingTransactions })
        },

        sendTransaction: async (
          maxPriorityFeePerGas: ethers.BigNumber,
          maxFeePerGas: ethers.BigNumber,
          gasLimit: number
        ) => {
          const { transaction } = get()
          const getCurrentAccount = useSettings.getState().currentAccount
          if (getCurrentAccount && transaction) {
            const currentAccountWallet = useStore.getState().walletsList[getCurrentAccount?.address]
            const encryptedPrivateKey = currentAccountWallet[transaction.asset.shortName]?.encryptedPrivateKey

            const fromAddress = transaction.asset.address

            const tranData = await NetworkFactory.selectByNetworkId(transaction.asset.shortName).sendTransaction({
              maxPriorityFeePerGas,
              maxFeePerGas,
              gasLimit,
              encryptedPrivateKey,
              transaction,
              fromAddress,
            })
            const { nonce, txHash, txDetails } = tranData
            set(
              produce((state: WalletState) => {
                if (state.transaction) {
                  state.transaction.hash = txHash
                  state.transaction.nonce = nonce || ''
                }
                if (txDetails) {
                  state.pendingTransactions?.unshift({
                    ...txDetails,
                    hash: txHash,
                    nonce,
                    value: state.transaction?.amount,
                  })
                }
              })
            )
            return tranData
          } else {
            throw new Error('Current Account Or Transaction not found')
          }
        },

        cancelTransaction: async (
          network: string,
          txData: {
            nonce: string
            maxPriorityFeePerGas: ethers.BigNumber
            maxFeePerGas: ethers.BigNumber
          }
        ) => {
          const { wallet } = get()
          if (wallet) {
            const provider = getNetworkProvider(network)
            const walletSigner = wallet.connect(provider)
            await walletSigner.sendTransaction({
              ...txData,
              data: '0x',
              to: ethers.constants.AddressZero,
              value: ethers.utils.parseEther('0.1'),
              gasLimit: 58000,
            })
          }
        },

        setLockTimer: (lockTime: number) => {
          chrome.storage.local.set({ LOCK_TIME: lockTime })
          set({ lockTime })
        },

        setLockWallet: (lockWallet: boolean) => {
          set({ lockWallet })
        },

        addCustomToken: async (tokenData: Asset) => {
          set(
            produce((state: WalletState) => {
              state.assets.push(tokenData)
            })
          )

          useSettings.getState().saveAccount()
        },

        removeToken: (network: string, assetId: string) => {
          set(
            produce((state: WalletState) => {
              state.assets = state.assets.filter((a) => !(a.id === assetId && a.network === network))
            })
          )

          useSettings.getState().saveAccount()
        },

        changeAvatar: (avatar: string) => {
          set(
            produce((state: WalletState) => {
              state.avatar = avatar
            })
          )

          useSettings.getState().saveAccount()
        },

        addNFT: (nft: Asset) => {
          set(
            produce((state: WalletState) => {
              state.NFTs.push(nft)
            })
          )

          useSettings.getState().saveAccount()
        },

        removeNFT: (nft: Asset) => {
          set(
            produce((state: WalletState) => {
              state.NFTs = state.NFTs.filter((i) => i.id !== nft.id && i.contractAddress === nft.contractAddress)
            })
          )

          useSettings.getState().saveAccount()
        },

        getNetworkList: () => {
          const { assets } = get()
          const assetsWithImage = assets.map(({ image, network }) => ({
            network,
            image,
          }))

          return [...new Map(assetsWithImage.map((item) => [item['network'], item])).values()]
        },

        setTokenProfitLoss: (data: TokenProfitCollection) => {
          const { tokenProfitLossCollection } = get()
          const token = tokenProfitLossCollection.find((v) => v.network === data.network && v.token === data.token)
          const tokenIndex = tokenProfitLossCollection.findIndex(
            (v) => v.network === data.network && v.token === data.token
          )
          const timeDifference = moment().diff(moment(data.time || token?.time), 'hours')
          if (!token) {
            set({
              tokenProfitLossCollection: [...tokenProfitLossCollection, data],
            })
          } else if (timeDifference >= 24) {
            // Replace value and Time to current
            set(
              produce((state: WalletState) => {
                state.tokenProfitLossCollection[tokenIndex].value = data.value
                state.tokenProfitLossCollection[tokenIndex].time = moment().toDate()
              })
            )
          }
        },
        setUsername: (username: string) => {
          set({
            username,
          })
        },
        setSelectedNetwork: (selectedNetwork: string) => {
          set({
            selectedNetwork: selectedNetwork,
          })
        },
      }),
      {
        name: 'wallet-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))),
      }
    ),
    { name: 'Shuttle - Wallet' }
  )
)
