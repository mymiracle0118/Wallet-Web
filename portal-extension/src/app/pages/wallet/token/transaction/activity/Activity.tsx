import React, { useEffect, useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ethers } from 'ethers'
import { useWallet } from '@portal/shared/hooks/useWallet'
import { createLocationState, goBack } from 'lib/woozie'
import ETH from 'assets/coins/ETH.svg'
import Alert from 'assets/icons/alert-triangle.svg'
import PasswordPromptModal from 'app/pages/wallet/PasswordPromptModal'
import { Button, COLORS, CustomTypography, Icon, useModalContext } from 'components'
import TransactionActivity from './TransactionActivity'
import { Modal, ModalBody, ModalContent } from '@nextui-org/react'
import { SpinnerIcon } from '@src/app/components/Icons'
import { NetworkFactory } from '@portal/shared/factory/network.factory'

const Activity = () => {
  const { t } = useTranslation()
  const [loading, setLoading] = useState<boolean>(false)
  const [openRequestModal, setOpenRequestModal] = useState<boolean>(false)
  const { address, cancelTransaction, wallet, setPendingTransactions, pendingTransactions } = useWallet() //,getAsset
  const [transaction, setTransaction] = useState<ethers.providers.TransactionResponse>()
  const [openPasswordModal, setOpenPasswordModal] = useState<boolean>(false)
  const [transactionStatus, setTransactionStatus] = useState('')
  const { setModalData } = useModalContext()

  const { pathname } = createLocationState()
  const paths = pathname.split('/')
  const network: string = paths[paths.length - 4]
  const assetId: string = paths[paths.length - 3]
  const transactionHash = paths[paths.length - 1]

  // const networkFactory = useMemo(() => NetworkFactory.selectNetwork(asset), [asset])
  const networkFactory = useMemo(() => NetworkFactory.selectByNetworkId(network))
  const fetchData = () => {
    if (transactionHash) {
      networkFactory
        .getTransaction(transactionHash)
        .then((tx: any) => {
          if (tx === null) {
            const removeCancelledTransaction = pendingTransactions?.filter((v) => v.hash !== transactionHash)
            setPendingTransactions(removeCancelledTransaction)
            goBack()
          } else {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            setTransaction(tx)
            if (tx.blockNumber) {
              setTransactionStatus('complete')
            }
          }
        })
        .catch((e) => console.log(e))
    }
  }

  useEffect(() => {
    fetchData()
    if (transactionStatus !== 'complete') {
      const pollInterval = setInterval(() => {
        fetchData()
        if (transactionStatus === 'complete') {
          clearInterval(pollInterval) // Stop polling when status is complete
        }
      }, 5000) // Poll every 5 seconds

      return () => {
        clearInterval(pollInterval) // Clean up interval when component unmounts
      }
    }
  }, [transactionStatus, transactionHash, network])

  const type = useMemo(
    () => (transaction?.from.toLowerCase() === address?.toLowerCase() ? 'Sent' : 'Received'),
    [address, transaction]
  )

  const handleCancelTransaction = async () => {
    if (wallet) {
      setLoading(true)
      const doubleGasFee = Number(ethers.utils.formatUnits(transaction?.gasPrice, 'gwei') * 10).toFixed(8)
      try {
        const { nonce } = transaction
        const tx = {
          nonce,
          maxPriorityFeePerGas: ethers.utils.parseUnits(`${doubleGasFee}`, 'gwei'),
          maxFeePerGas: ethers.utils.parseUnits(`${doubleGasFee}`, 'gwei'),
        }
        await cancelTransaction(network, tx)
        setLoading(false)
        setOpenRequestModal(false)
        setModalData({
          type: 'send-cancel',
          amount: ethers.utils.formatEther(transaction?.value),
          token: 'ETH',
          tokenImage: <ETH />,
        })
      } catch (err) {
        setLoading(false)
        console.error('Error: ', err)
      }
    } else {
      setOpenPasswordModal(true)
    }
  }

  const handlePasswordPromptSuccess = async () => {
    setOpenPasswordModal(false)
    await handleCancelTransaction()
  }
  return (
    <>
      <TransactionActivity
        page="Token"
        disableClick
        type={type}
        txStatus={
          transaction
            ? transaction.status
              ? transaction.status === 'Invalid'
                ? 'Failed'
                : transaction.status === 'Success'
                ? 'Completed'
                : 'Failed'
              : transaction.blockNumber
              ? 'Completed'
              : 'Pending'
            : ''
        }
        openModal={() => setOpenRequestModal(true)}
        network={network}
        assetId={assetId}
        transactionHash={transactionHash}
      />

      <Modal
        backdrop="opaque"
        isOpen={openRequestModal}
        onClose={() => setOpenRequestModal(false)}
        hideCloseButton={true}
        placement="center"
        className="max-w-[20.625rem]"
      >
        <ModalContent className="py-3 rounded-[1.75rem] dark:bg-surface-dark bg-custom-white">
          <ModalBody>
            <div className="flex flex-col items-center justify-center space-y-4">
              <div
                className="cursor-pointer rounded-full mx-auto text-[2.5rem] h-16 w-16 flex items-center justify-center"
                style={{ background: COLORS.background.gradientLogoBg }}
              >
                <Icon icon={<Alert />} size="inherit" />
              </div>
              <CustomTypography variant="h1" className="text-center">
                {t('Token.cancelTransaction')}
              </CustomTypography>
              <CustomTypography variant="subtitle" className="text-center dark:text-custom-white40">
                {t('Token.cancelTransactionModalInstruction')}
              </CustomTypography>
            </div>
            <div className="bg-surface-dark-alt p-4 rounded-lg space-y-3">
              <div className="flex items-start justify-between">
                <CustomTypography variant="subtitle">{t('Token.estimatedGasFee')}</CustomTypography>
                <div className="text-right">
                  <CustomTypography variant="subtitle">0.00437 ETH</CustomTypography>
                  <CustomTypography variant="body" color="text-custom-white40">
                    $14
                  </CustomTypography>
                </div>
              </div>
              <div className="flex items-center justify-between">
                <CustomTypography variant="subtitle">{t('Token.estimatedTime')}</CustomTypography>
                <CustomTypography variant="subtitle">{`< 30 seconds`}</CustomTypography>
              </div>
            </div>
            <div className="flex justify-between flex-row gap-2 pt-2">
              <Button variant="bordered" color="outlined" onClick={() => setOpenRequestModal(false)}>
                {t('Actions.Nothanks')}
              </Button>
              <Button
                onClick={handleCancelTransaction}
                color={`${loading ? 'disabled' : 'negative'}`}
                isDisabled={loading}
                isLoading={loading}
                spinner={<SpinnerIcon />}
              >
                {!loading && t('Actions.request')}
              </Button>
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>

      <PasswordPromptModal
        modalState={openPasswordModal}
        closePromptModal={() => setOpenPasswordModal(false)}
        onSuccess={handlePasswordPromptSuccess}
      />
    </>
  )
}

export default Activity
