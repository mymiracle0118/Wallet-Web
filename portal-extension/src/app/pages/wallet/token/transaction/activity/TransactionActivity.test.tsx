import React from 'react'
import { render, renderHook, act } from '@testing-library/react'
import { I18nextProvider } from 'react-i18next'
import i18n from '@portal/shared/i18n/config'
import Activity from 'pages/wallet/token/transaction/activity/Activity'
import { useWallet } from '@portal/shared/hooks/useWallet'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

const queryClient = new QueryClient()
const wrapper = (
  <QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
      <Activity />
    </I18nextProvider>
  </QueryClientProvider>
)

describe('Transaction Activity UI', () => {
  const originalLocation = window.location as URL
  const mockWindowLocation = (newLocation: URL) => {
    delete window.location
    window.location = newLocation
  }
  const setLocation = (path: string) => mockWindowLocation(new URL(`https://example.com${path}`))

  beforeEach(() => {
    const { result } = renderHook(() => useWallet())

    act(() => {
      result.current.network = 'goerli'
      result.current.address = '0x63C100ac0C36549C7218070294b60C18d813675c'
      setLocation('#/token/goerli/activity/0x7a0c1749b3f55d0be14f38c406fbcae08094bf2a3079e1c459a819d1d57d916e')
    })
  })

  afterEach(() => {
    // Restore window.location to not destroy other tests
    mockWindowLocation(originalLocation)
  })

  test('renders correctly', () => {
    const tree = render(wrapper)
    expect(tree).toMatchSnapshot()
  })
})
