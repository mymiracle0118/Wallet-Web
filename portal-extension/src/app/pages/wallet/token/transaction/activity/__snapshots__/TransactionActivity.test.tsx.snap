// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Activity UI renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
      >
        <div
          class="flex flex-row relative justify-between items-center h-16"
        >
          <button
            class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
            type="button"
          >
            <arrow-left.svg
              classname="text-xl stroke-custom-black dark:stroke-custom-white"
            />
          </button>
          <div
            class="flex-1 text-center"
          >
            <div
              class="text-custom-black dark:text-custom-white text-lg font-extrabold"
            >
              Activity
            </div>
          </div>
        </div>
        <div
          class="w-full h-full overflow-auto"
        >
          <div
            class="MuiBox-root css-0"
          />
        </div>
      </div>
      <div
        style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
      />
      <div
        style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
      />
    </div>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      >
        <div />
      </div>
      <div
        data-headlessui-portal=""
      >
        <div />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
    >
      <div
        class="flex flex-row relative justify-between items-center h-16"
      >
        <button
          class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
          type="button"
        >
          <arrow-left.svg
            classname="text-xl stroke-custom-black dark:stroke-custom-white"
          />
        </button>
        <div
          class="flex-1 text-center"
        >
          <div
            class="text-custom-black dark:text-custom-white text-lg font-extrabold"
          >
            Activity
          </div>
        </div>
      </div>
      <div
        class="w-full h-full overflow-auto"
      >
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
    <div
      style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
    <div
      style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
